src/Microsoft.Graph.Core/Requests/RedirectHandler.cs

```diff
@@ -0,0 +1,159 @@
+    public class RedirectHandler : DelegatingHandler
+    {
+
+        private const int maxRedirects = 5;
+
+        /// <summary>
+        /// Constructs a new <see cref="RedirectHandler"/> 
+        /// </summary>
+        /// <param name="innerHandler">An HTTP message handler to pass to the <see cref="HttpMessageHandler"/> for sending requests.</param>
+        public RedirectHandler(HttpMessageHandler innerHandler)
+        {
+            InnerHandler = innerHandler;
+        }
+
+        /// <summary>
+        /// Disposes the HttpMessageHandler instance.
+        /// </summary>
+        public new void Dispose()
+        {
+            if (InnerHandler != null)
+            {
+                InnerHandler.Dispose();
+            }
+
+        }
+
+        /// <summary>
+        /// Sends the Request 
+        /// </summary>
+        /// <param name="request">The <see cref="HttpRequestMessage"/> to send.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/>for the request.</param>
+        /// <returns>The <see cref="HttpResponseMessage"/>.</returns>
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            
+            // send request first time to get response
+            var response = await base.SendAsync(request, cancellationToken);
+
+            // check response status code 
+            if (IsRedirect(response.StatusCode))
+            {
+
+                // general copy request with internal copyRequest(see copyRequest for details) Method 
+                var newRequest = CopyRequest(request);
+                StreamContent content = null;
+
+                if (request.Content != null && request.Content.Headers.ContentLength != 0)
+                {
+                    content = new StreamContent(request.Content.ReadAsStreamAsync().Result);
+                }
+
+                var redirectCount = 0;
+
+                // check whether redirect count over maxRedirects
+                while (redirectCount < maxRedirects)
+                {
+                   
+                    // status code == 303: change request method from post to get and content to be null
+                    if (response.StatusCode == HttpStatusCode.SeeOther)
+                    {
+                        newRequest.Content = null;
+                        newRequest.Method = HttpMethod.Get;
+                    }
+                    else
+                    {
+                        newRequest.Content = content;
+                        newRequest.Method = request.Method;
+                    }
+
+                    // Set newRequestUri from response
+                    newRequest.RequestUri = response.Headers.Location;
+
+                    // Remove Auth if unneccessary
+                    if (String.Compare(newRequest.RequestUri.Host, request.RequestUri.Host, StringComparison.OrdinalIgnoreCase) != 0)
```

Comment:
We need to check the schema as well.  e.g. If the original is https and it redirects to http then the auth header must be dropped also.

Reply:
Thanks. will update it.